File: 46-permutations/46-permutations.cpp
class Solution {
 public:   
    void Track(vector<vector<int>>& res, vector<int>& nums, vector<int>& permutation, vector<bool>& used){
     if(permutation.size() == nums.size()){
         res.push_back(permutation);
         return; 
     }
        
        for(int i=0;i<nums.size();i++){
            if (!used[i]){     
                //valid option
                used[i]= true;
                permutation.push_back(nums[i]);
                Track(res, nums, permutation, used);
                //Undo option
                used[i]= false;
                permutation.pop_back();
                    
            }
        }
    }
    

    vector<vector<int>> permute(vector<int>& nums) {
        vector<vector<int>> res;
        vector<bool> used(nums.size(),false);
        vector<int> permutation;
        Track(res, nums, permutation, used);
        return res;
        
    }
};

File: 46-permutations/46-permutations.java
class Solution {
        public List<List<Integer>> permute(int[] nums) {

        List<List<Integer>> res =new ArrayList<>();

     HashSet<Integer> hashset = IntStream.of(nums).boxed().collect(Collectors.toCollection(HashSet::new));


        permuteRec(hashset, new ArrayList<Integer>() ,res );
      return res;

    }

    public void permuteRec(HashSet <Integer>nums, List<Integer> prefix, List<List<Integer>> res){
            if(nums.size()==0){
               res.add(new ArrayList<>(prefix));
                return;
            }
            HashSet<Integer> remainingnums= new HashSet<Integer>(nums);
            for(int value: nums){
                remainingnums.remove(value);
                prefix.add(value);
                permuteRec(remainingnums,prefix,res);
                remainingnums.add(value);
                prefix.remove(prefix.size()-1);
            }
    }
}